File: ai-service/app/utils/clean_logger.py
"""
Clean and concise logger with aggregation support
Provides readable, summarized output with minimal noise
"""

import os
import time
from typing import Dict, Any, Optional
from collections import Counter


class Colors:
    """ANSI color codes"""
    RESET = '\033[0m'
    BOLD = '\033[1m'
    GRAY = '\033[90m'
    RED = '\033[31m'
    GREEN = '\033[32m'
    YELLOW = '\033[33m'
    BLUE = '\033[34m'
    CYAN = '\033[36m'
    WHITE = '\033[37m'


class CleanLogger:
    """Clean, aggregated logger for AI service"""

    def __init__(self):
        self.is_verbose = os.getenv('LOG_VERBOSE', 'false').lower() == 'true'

    def _color(self, text: str, color: str) -> str:
        """Apply color to text"""
        return f"{color}{text}{Colors.RESET}"

    def endpoint(self, method: str, path: str, meta: Optional[Dict[str, Any]] = None):
        """Log an API endpoint call"""
        timestamp = time.strftime('%H:%M:%S')
        print(f"\n{self._color('►', Colors.CYAN)} {self._color(Colors.BOLD + method + Colors.RESET, Colors.WHITE)} {path} {self._color(f'[{timestamp}]', Colors.GRAY)}")

        if meta:
            for key, value in meta.items():
                print(f"  {self._color('│', Colors.GRAY)} {key}: {self._color(str(value), Colors.WHITE)}")

    def step(self, step_name: str, details: Optional[Dict[str, Any]] = None):
        """Log a processing step"""
        print(f"  {self._color('●', Colors.BLUE)} {self._color(Colors.BOLD + step_name + Colors.RESET, Colors.WHITE)}")

        if details:
            for key, value in details.items():
                print(f"    {self._color('•', Colors.GRAY)} {key}: {self._color(str(value), Colors.WHITE)}")

    def success(self, message: str, summary: Optional[Dict[str, Any]] = None):
        """Log a success result"""
        print(f"  {self._color('✓', Colors.GREEN)} {self._color(Colors.BOLD + message + Colors.RESET, Colors.GREEN)}")

        if summary:
            for key, value in summary.items():
                print(f"    {self._color('•', Colors.GRAY)} {key}: {self._color(str(value), Colors.GREEN)}")

    def error(self, message: str, error: Optional[Exception] = None):
        """Log an error"""
        print(f"  {self._color('✗', Colors.RED)} {self._color(Colors.BOLD + message + Colors.RESET, Colors.RED)}")

        if error and self.is_verbose:
            print(f"    {self._color('•', Colors.GRAY)} {self._color(str(error), Colors.RED)}")

    def warn(self, message: str, details: Optional[str] = None):
        """Log a warning"""
        print(f"  {self._color('⚠', Colors.YELLOW)} {self._color(message, Colors.YELLOW)}")

        if details and self.is_verbose:
            print(f"    {self._color('•', Colors.GRAY)} {self._color(details, Colors.GRAY)}")

    def aggregated(self, title: str, items: Dict[str, int], show_total: bool = True, max_items: int = 10):
        """Log aggregated results with visual bars"""
        print(f"\n  {self._color('◆', Colors.CYAN)} {self._color(Colors.BOLD + title + Colors.RESET, Colors.WHITE)}")

        entries = list(items.items())
        total = sum(count for _, count in entries if isinstance(count, int))

        # Sort by count (descending) and limit
        sorted_entries = sorted(entries, key=lambda x: x[1] if isinstance(x[1], int) else 0, reverse=True)[:max_items]

        for key, value in sorted_entries:
            bar = self._create_bar(value, total)
            print(f"    {self._color('│', Colors.GRAY)} {self._color(str(key).ljust(30), Colors.WHITE)} {self._color(str(value), Colors.CYAN)} {bar}")

        if len(entries) > max_items:
            print(f"    {self._color('│', Colors.GRAY)} {self._color(f'... and {len(entries) - max_items} more', Colors.GRAY)}")

        if show_total and len(entries) > 1:
            print(f"    {self._color('│', Colors.GRAY)} {self._color(Colors.BOLD + 'TOTAL'.ljust(30) + Colors.RESET, Colors.WHITE)} {self._color(Colors.BOLD + str(total) + Colors.RESET, Colors.CYAN)}")

    def _create_bar(self, value: int, total: int, max_width: int = 15) -> str:
        """Create a text-based progress bar"""
        if not total or total == 0:
            return ''

        percentage = min(value / total, 1.0)
        filled = round(percentage * max_width)
        empty = max_width - filled

        return (self._color('[', Colors.GRAY) +
                self._color('█' * filled, Colors.CYAN) +
                self._color('░' * empty, Colors.GRAY) +
                self._color(']', Colors.GRAY))

    def debug(self, message: str, data: Any = None):
        """Log debug information (only if LOG_VERBOSE=true)"""
        if not self.is_verbose:
            return

        print(f"  {self._color('◯', Colors.GRAY)} {self._color(message, Colors.GRAY)}")
        if data:
            print(f"    {self._color(str(data), Colors.GRAY)}")

    def separator(self):
        """Log a separator line"""
        print(f"{self._color('─' * 60, Colors.GRAY)}")

    def start_timer(self) -> float:
        """Start timing an operation"""
        return time.time()

    def get_elapsed(self, start_time: float) -> str:
        """Get elapsed time in human-readable format"""
        elapsed = time.time() - start_time

        if elapsed < 1:
            return f"{int(elapsed * 1000)}ms"
        elif elapsed < 60:
            return f"{elapsed:.1f}s"
        else:
            minutes = int(elapsed / 60)
            seconds = int(elapsed % 60)
            return f"{minutes}m {seconds}s"


# Export singleton instance
clean_logger = CleanLogger()

File: backend/src/utils/clean-logger.js
/**
 * Clean and concise logger with aggregation support
 * Provides readable, summarized output with minimal noise
 */

class CleanLogger {
  constructor() {
    this.isVerbose = process.env.LOG_VERBOSE === 'true';

    // ANSI color codes
    this.colors = {
      reset: '\x1b[0m',
      bold: '\x1b[1m',
      gray: '\x1b[90m',
      red: '\x1b[31m',
      green: '\x1b[32m',
      yellow: '\x1b[33m',
      blue: '\x1b[34m',
      cyan: '\x1b[36m',
      white: '\x1b[37m',
    };
  }

  _color(text, color) {
    return `${this.colors[color]}${text}${this.colors.reset}`;
  }

  /**
   * Log an API endpoint call
   */
  endpoint(method, path, { ...meta } = {}) {
    const timestamp = new Date().toLocaleTimeString();
    console.log(`\n${this._color('►', 'cyan')} ${this._color(this.colors.bold + method + this.colors.reset, 'white')} ${path} ${this._color(`[${timestamp}]`, 'gray')}`);

    if (Object.keys(meta).length > 0) {
      Object.entries(meta).forEach(([key, value]) => {
        console.log(`  ${this._color('│', 'gray')} ${key}: ${this._color(value, 'white')}`);
      });
    }
  }

  /**
   * Log a processing step with clean formatting
   */
  step(stepName, details = {}) {
    console.log(`  ${this._color('●', 'blue')} ${this._color(this.colors.bold + stepName + this.colors.reset, 'white')}`);

    Object.entries(details).forEach(([key, value]) => {
      console.log(`    ${this._color('•', 'gray')} ${key}: ${this._color(value, 'white')}`);
    });
  }

  /**
   * Log a success result with summary
   */
  success(message, summary = {}) {
    console.log(`  ${this._color('✓', 'green')} ${this._color(this.colors.bold + message + this.colors.reset, 'green')}`);

    if (Object.keys(summary).length > 0) {
      Object.entries(summary).forEach(([key, value]) => {
        console.log(`    ${this._color('•', 'gray')} ${key}: ${this._color(value, 'green')}`);
      });
    }
  }

  /**
   * Log an error with context
   */
  error(message, error = null) {
    console.log(`  ${this._color('✗', 'red')} ${this._color(this.colors.bold + message + this.colors.reset, 'red')}`);

    if (error && this.isVerbose) {
      console.log(`    ${this._color('•', 'gray')} ${this._color(error.message, 'red')}`);
      if (error.stack) {
        console.log(`    ${this._color(error.stack.split('\n').slice(0, 3).join('\n    '), 'gray')}`);
      }
    }
  }

  /**
   * Log a warning
   */
  warn(message, details = null) {
    console.log(`  ${this._color('⚠', 'yellow')} ${this._color(message, 'yellow')}`);

    if (details && this.isVerbose) {
      console.log(`    ${this._color('•', 'gray')} ${this._color(details, 'gray')}`);
    }
  }

  /**
   * Log aggregated results (e.g., deviation counts by type)
   */
  aggregated(title, items, { showTotal = true, maxItems = 10 } = {}) {
    console.log(`\n  ${this._color('◆', 'cyan')} ${this._color(this.colors.bold + title + this.colors.reset, 'white')}`);

    const entries = Object.entries(items);
    const total = entries.reduce((sum, [, count]) => sum + (typeof count === 'number' ? count : 0), 0);

    const sortedEntries = entries
      .sort((a, b) => (typeof b[1] === 'number' && typeof a[1] === 'number') ? b[1] - a[1] : 0)
      .slice(0, maxItems);

    sortedEntries.forEach(([key, value]) => {
      const bar = this._createBar(value, total);
      console.log(`    ${this._color('│', 'gray')} ${this._color(key.padEnd(25), 'white')} ${this._color(value, 'cyan')} ${bar}`);
    });

    if (entries.length > maxItems) {
      console.log(`    ${this._color('│', 'gray')} ${this._color(`... and ${entries.length - maxItems} more`, 'gray')}`);
    }

    if (showTotal && entries.length > 1) {
      console.log(`    ${this._color('│', 'gray')} ${this._color(this.colors.bold + 'TOTAL'.padEnd(25) + this.colors.reset, 'white')} ${this._color(this.colors.bold + total + this.colors.reset, 'cyan')}`);
    }
  }

  /**
   * Create a simple text-based progress bar
   */
  _createBar(value, total, maxWidth = 15) {
    if (!total || total === 0) return '';

    const percentage = Math.min(value / total, 1);
    const filled = Math.round(percentage * maxWidth);
    const empty = maxWidth - filled;

    return this._color('[', 'gray') +
           this._color('█'.repeat(filled), 'cyan') +
           this._color('░'.repeat(empty), 'gray') +
           this._color(']', 'gray');
  }

  /**
   * Log verbose debug information (only if LOG_VERBOSE=true)
   */
  debug(message, data = null) {
    if (!this.isVerbose) return;

    console.log(`  ${this._color('◯', 'gray')} ${this._color(message, 'gray')}`);
    if (data) {
      console.log(`    ${this._color(JSON.stringify(data, null, 2), 'gray')}`);
    }
  }

  /**
   * Log a clean separator
   */
  separator() {
    console.log(`${this._color('─'.repeat(60), 'gray')}`);
  }

  /**
   * Start timing an operation
   */
  startTimer() {
    return Date.now();
  }

  /**
   * Get elapsed time in human-readable format
   */
  getElapsed(startTime) {
    const elapsed = Date.now() - startTime;

    if (elapsed < 1000) return `${elapsed}ms`;
    if (elapsed < 60000) return `${(elapsed / 1000).toFixed(1)}s`;
    return `${Math.floor(elapsed / 60000)}m ${Math.floor((elapsed % 60000) / 1000)}s`;
  }
}

// Export singleton instance
const cleanLogger = new CleanLogger();

module.exports = cleanLogger;
